webknossos/webknossos/dataset/storage.py

```diff
@@ -0,0 +1,408 @@
+import re
+import warnings
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from os.path import relpath
+from pathlib import Path
+from typing import Any, Dict, Iterator, Optional, Type
+
+import numpy as np
+import wkw
+import zarr
+from numcodecs import Blosc
+
+from webknossos.geometry import BoundingBox, Vec3Int, Vec3IntLike
+
+
+def is_power_of_two(num: int) -> bool:
+    return num & (num - 1) == 0
+
+
+class StorageArrayException(Exception):
+    pass
+
+
+class StorageArrayFormat(Enum):
+    WKW = "wkw"
+    Zarr = "zarr"
+
+
+@dataclass
+class StorageArrayInfo:
+    array_format: StorageArrayFormat
+    num_channels: int
+    voxel_type: np.dtype
+    chunk_size: Vec3Int
+    chunks_per_shard: Vec3Int
+    compression_mode: bool = False
+
+    @property
+    def shard_size(self) -> Vec3Int:
+        return self.chunk_size * self.chunks_per_shard
+
+
+class StorageArray(ABC):
```

Comment:
I think it `StorageArray` is not too far-fetched, because this layer of abstraction is supposed to be the closest layer to the actual persistance in this library. But, we can call it `Array` or `BaseArray`. Or even `DataArray` to fit the `data_format` attribute :trollface: